apollo-entity/src/main/java/com/spotify/apollo/entity/CodecEntityMiddleware.java

```diff
@@ -19,79 +19,73 @@
  */
 package com.spotify.apollo.entity;
 
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
 import com.spotify.apollo.RequestContext;
 import com.spotify.apollo.Response;
 import com.spotify.apollo.Status;
 import com.spotify.apollo.route.AsyncHandler;
 import com.spotify.apollo.route.Middleware;
 import com.spotify.apollo.route.SyncHandler;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import java.io.IOException;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.function.Function;
-
 import javaslang.control.Either;
 import okio.ByteString;
-
-import static java.util.concurrent.CompletableFuture.completedFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * A factory for creating middlewares that can be used to create routes that deal directly
  * with an api entity.
  *
- * A {@link EntityCodec} is used to define how to go between a {@link ByteString} and the
- * entity type used in your route handlers.
+ * A {@link Codec} is used to define how to go between a {@link ByteString} and the entity type
+ * used in your route handlers.
  */
 class CodecEntityMiddleware implements EntityMiddleware {
 
   private static final Logger LOG = LoggerFactory.getLogger(CodecEntityMiddleware.class);
 
   private static final String CONTENT_TYPE = "Content-Type";
 
-  private final EntityCodec codec;
-  private final String contentType;
-
-  CodecEntityMiddleware(EntityCodec codec) {
-    this(codec, codec.defaultContentType());
-  }
+  private final Codec codec;
 
-  CodecEntityMiddleware(EntityCodec codec, String contentType) {
+  CodecEntityMiddleware(Codec codec) {
     this.codec = Objects.requireNonNull(codec);
-    this.contentType = Objects.requireNonNull(contentType);
   }
 
   @Override
   public <R> Middleware<SyncHandler<R>, SyncHandler<Response<ByteString>>>
   serializerDirect(Class<? extends R> responseEntityClass) {
-    return inner -> inner
+    return inner -> rc -> inner
         .map(Response::forPayload)
-        .map(serialize(responseEntityClass));
+        .map(serialize(rc, responseEntityClass))
+        .invoke(rc);
```

Comment:
Wow. That's kind of mind-boggling. Is there a more obvious way to write this code, maybe?

Reply:
I know it's a bit dense. I tried expanding it a bit but I have a hard time finding meaningful names for the intermediate handlers. Here's an example:

```java
@Override
public <R> Middleware<SyncHandler<R>, SyncHandler<Response<ByteString>>>
    serializerDirect(Class<? extends R> responseEntityClass) {
  return inner -> {
    final SyncHandler<Response<ByteString>> wrappedHandler = rc -> {
      final SyncHandler<Response<ByteString>> serializingHandler = inner
          .map(Response::forPayload)
          .map(serialize(rc, responseEntityClass));

      return serializingHandler.invoke(rc);
    };
    return wrappedHandler;
  };
}
```

or this one that is slightly less expanded:

```java
@Override
public <R> Middleware<SyncHandler<R>, SyncHandler<Response<ByteString>>>
    serializerDirect(Class<? extends R> responseEntityClass) {
  return inner -> rc -> {
    final SyncHandler<Response<ByteString>> serializingHandler = inner
        .map(Response::forPayload)
        .map(serialize(rc, responseEntityClass));

    return serializingHandler.invoke(rc);
  };
}
```

Reply:
I like the second option, that is easier for me to read.

Reply:
Refactored 👍